server/src/main/java/org/opensearch/gateway/TransportNodesListGatewayStartedShardsBatch.java

```diff
@@ -136,27 +135,36 @@ protected NodesGatewayStartedShardsBatch newResponse(
     @Override
     protected NodeGatewayStartedShardsBatch nodeOperation(NodeRequest request) {
         Map<ShardId, GatewayStartedShard> shardsOnNode = new HashMap<>();
-        for (ShardAttributes shardAttr : request.shardAttributes.values()) {
-            final ShardId shardId = shardAttr.getShardId();
+        for (Map.Entry<ShardId, ShardAttributes> shardAttr : request.shardAttributes.entrySet()) {
+            final ShardId shardId = shardAttr.getKey();
             try {
                 shardsOnNode.put(
                     shardId,
-                    getShardInfoOnLocalNode(
-                        logger,
-                        shardId,
-                        namedXContentRegistry,
-                        nodeEnv,
-                        indicesService,
-                        shardAttr.getCustomDataPath(),
-                        settings,
-                        clusterService
+                    new GatewayStartedShard(
```

Comment:
you need to merge from main, this has been merged already

Reply:
TransportNodesListGatewayStartedShardsBatch.GatewayStartedShard is different than the helper one. (I can rename it if needed, just following the pattern which got pushed in Replica batch class)

This GatewayStartedShard contains shard metadata and transport level exception if thrown from code for a shard.

```
Helper.GatewayStartedShard 
Exception
```

This exception is used to differentiate between `Helper.GatewayStartedShard.storeException` and actual code level exceptions.

**Helper.GatewayStartedShard.storeException** = store level exception happened during load shard path etc, which can be used in decision making
**Exception transportError**  = used to decide if we need to retry the shard or ignore based on non retry-able exceptions

Similar to Replica batch.

Reply:
> This GatewayStartedShard contains shard metadata and transport level exception if thrown from code for a shard.

why there would be transport exception? There can be store exception when reading the shards on node. TransportException would be for the whole batch right?

Reply:
Not exactly "Transport Exception" but some other exception from rest of the code which are not handled. Handling of store related unknown exception and rest of the code related unknown exceptions **are different**. 

* store related unknown exception => Save in cache and give exception to allocator
* rest of the code related unknown exceptions => Ignore response and don't give node's exception to allocator

### Existing Handling of Exceptions
For example - In primary transport we can get 3 types of exceptions 
1. [Index not found](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/gateway/TransportNodesGatewayStartedShardHelper.java#L75)
2. [No shard path](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/gateway/TransportNodesGatewayStartedShardHelper.java#L83)
3. [Exception while opening the index](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/gateway/TransportNodesGatewayStartedShardHelper.java#L98) - this is part of Exception catch block so we don't know the type. And it's passed within the response object. [Unknown one]

Now 1st and 2nd is thrown from the transport and it is received [on the failure handling flow.](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/gateway/AsyncShardFetchCache.java#L195) We ignore the exception/failure at the cache layer and don't use its details when data is returned via cache. Before ignoring the response, we also check [if the exception is retry-able](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/gateway/AsyncShardFetchCache.java#L208). Now we can see that the exceptions which are retry-able are at request level only, but we still check it if any other part of the code throws any exception other than 1st and 2nd.  
* OpenSearchRejectedExecutionException
* ReceiveTimeoutTransportException
* OpenSearchTimeoutException
But we still check if the exception thrown from transport is retry-able and then only ignore it. 

Now another flow which can throw an `IOException` from transport is [ShardStateMetadata.FORMAT.loadLatestState](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/gateway/TransportNodesGatewayStartedShardHelper.java#L56-L60) -> this is not handled in code but it can still throw.
Today in case of this exception, first we unwrap the exception, check if it's retry-able then ignore its response most probably because it'll not be part of those 3 retry-able ones mentioned above.

Now the third exception(Exception while opening the index) is passed within the object and it's received on the [response layer](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/gateway/AsyncShardFetchCache.java#L164) and it's used in Allocators, because we store it in the cache as it is. 


### New Flow

In ShardBatchCache right now, I'm doing the handling of "**Rest of the code related unknown exceptions thrown from one shard execution**". 
Handling is same as before to maintain the behaviour. 
If exception is non-retryable - we'll ignore the response of the shard. 
If we use the `storeException` parameter of **GatewayStartedShard** , we'll miss the exception of "Exception while opening the index" (3rd case) and we'll ignore it. To avoid that I differentiated **storeException** and "**rest of the code related exceptions**"

## To use only single exception parameter

We need to assume that 
* The exceptions written as part of retry-able ones will never be thrown from the shard level code flow executions, not even from [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/gateway/TransportNodesGatewayStartedShardHelper.java#L98)
* I'll remove the handling of [retry-able exceptions from shard level](https://github.com/opensearch-project/OpenSearch/pull/12504/files#diff-f22c9cb55b850db5c383ce5ebfa022765e7089d220e80bd216b208cc0e7864fcR228-R251) responses as we just assumed it can never come.
* Store any exception which comes as part of the response as it is in the cache, because we need it. [[One specific case](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/gateway/TransportNodesGatewayStartedShardHelper.java#L98) of primary]
* Send null in response for other code related exceptions for example **IOException**, as we anyway intend to just ignore it. So now we need to start sending null similar to other exceptions like [IllegalStateException](https://github.com/opensearch-project/OpenSearch/pull/12504/files#diff-19afc691cf39869946cc46e71b0ab6aaec7373f563f8b4098dec6808090797fbR159-R161) or [INDEX_NOT_FOUND](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java#L158)

Changes in Primary and Replica transport would include 

Primary
```
if (e instanceof IllegalStateException || e instanceof IOException || e.getMessage().contains(INDEX_NOT_FOUND)) {
                    shardsOnNode.put(shardId, null);
                } 
```
Replica
```
if (e.getMessage().contains(INDEX_NOT_FOUND) || e instanceof IOException) {
                    shardStoreMetadataMap.put(shardId, null);
                }
```

Today, IOException is the only exception which is being thrown for single shard level execution, so we'll handle that. 
Benefits :
1. We can avoid adding any new exception param for "rest of the code related exceptions" in Primary transport. Though, in replica we've already added because there was none before.
2. `ShardBatchCache` putData function will be simplified as it does not need to filter any shards, it'll just store the data received. 

@shwetathareja  - please let me know your thoughts on this. This will require some testing, I'll work on that.

Reply:
why we need to make assumption? What happens in single shard flow when a new runtime exception occur, maintain the same logic here as well.

Reply:
Added code comments that those 3 retryable exceptions can't be thrown from the for loop we're executing. If any changes are being done in that flow, exception handling should be re-visited. 

I'll send null for IllegalState, Index not found and IOException as we do the same today. For any other exception, it'll be stored in cache.